pkg-manager/core/src/install/index.ts

```diff
@@ -1223,6 +1237,14 @@ const _installInContext: InstallFunction = async (projects, ctx, opts) => {
     strictPeerDependencies: opts.strictPeerDependencies,
   })
 
+  // Similar to the sequencing for when the original wanted lockfile is
+  // copied, the new lockfile passed here should be as close as possible to
+  // what will eventually be written to disk. Ex: peers should be resolved,
+  // the afterAllResolved hook has been applied, etc.
+  if (originalLockfileForCheck != null) {
+    opts.lockfileCheck?.(originalLockfileForCheck, newLockfile)
```

Comment:
why should this be a hook? Can't we just return the lockfile from the function?

Reply:
We could propagate the `newLockfile` all the way up to the `installDeps` function. That would involve more refactoring, but it would work.

I opted for the hook since we'd still need to pass down a `lockfileCheckOnly` (or similarly named) boolean, and it felt right to just pass down `lockfileCheck` instead and let that assume `lockfileCheckOnly`.

We'd also need to pass up a nullable `originalLockfileForCheck` through several functions, which feels a bit error prone since it can be nullable. There's a possibility other code relies on `originalLockfileForCheck` along the call hierarchy in the future and it may not be obvious that return value depends `lockfileCheckOnly`.

Reply:
Maybe it should be refactored then. Maybe there needs to be a new function. After all, it doesn't look like all the features of mutateModules are needed for this.

Reply:
There likely is a refactor that would clean up the install code a lot. I had noticed creating the new lockfile and mutating modules are operations that are really tied together.

How would you feel about a refactor that splits that into two separate phases?

1. Part one that computes the next wanted lockfile and returns it.
2. Part two that takes the new wanted lockfile as input and mutates modules.

Or did you mean something different by "_Maybe there needs to be a new function._"?

Reply:
The refactor in the comment above would be difficult to get right without accidentally introducing behavior changes since there's shared `ctx` between computing the next wanted lockfile and mutating modules. I do see a lot of value in that refactor outside of making it easier to `pnpm dedupe --check` though.

Reply:
The `opts.lockfileOnly` option is in a similarly strange state where we check that a bunch of times after computing the next lockfile to skip subsequent operations like mutating modules.

Reply:
@zkochan Let me know if you still want a refactor here. Just want to agree on some specifics before I go ahead.

Reply:
a refactor would be good. I don't know how it should be done now.